src/main/resources/assets/admin/common/js/form/set/FormSetOccurrenceView.ts

```diff
@@ -436,21 +439,67 @@ export abstract class FormSetOccurrenceView
 
                 if (label.length > 0) {
                     labels.push(label);
-
-                    if (firstOnly) {
-                        return true;
-                    }
                 }
             } else if (ValueTypes.DATA.equals(prop.getType())) {
-                return prop.getPropertySet().getPropertyArrays().some(arr => {
-                    this.recursiveFetchLabels(arr, labels, firstOnly);
-                    return firstOnly && labels.length > 0;
-                });
+                const propertySet = prop.getPropertySet();
+                const formItem = this.getFormItemByName(prop.getName());
+                if (formItem instanceof FormOptionSet) {
+                    const selectionArray = propertySet.getPropertyArray('_selected');
+
+                    if (selectionArray && !selectionArray.isEmpty()) {
+                        const nameLabelMap = new Map<string, any>(
+                            formItem.getFormItems()
+                                .filter((fi: FormItem) => fi instanceof FormOptionSetOption)
+                                .map((fo: FormOptionSetOption, index: number) => {
+                                    return [fo.getName(), {label: fo.getLabel(), index: index}] as [string, any];
+                                })
+                        );
+
+                        const selectedLabels = selectionArray.getProperties()
+                            .sort((one: Property, two: Property) => {
+                                return nameLabelMap.get(one.getString()).index - nameLabelMap.get(two.getString()).index;
+                            })
+                            .map((selectedProp: Property) => nameLabelMap.get(selectedProp.getString()).label);
+
+                        labels.push(...selectedLabels);
+
+                        if (labels.length === 0) {
+                            // should not happen, but in case no labels were found go inside selected options forms
+                            selectionArray.some((selectedProp) => {
```

Comment:
Can we please refactor this?

```Typescript
if (formItem instanceof FormOptionSet) {
  this.fetchLabelFromOptionSet(propertySet);
elseif (formItem instanceof FormItemSet) {
  this.fetchLabelFromItemSet(propertySet);
} else {
  ...
}
```

It actually seems like in the end you always do

```Typescript
this.recursiveFetchLabels(arr, labels, firstOnly);
return firstOnly && labels.length > 0;

```
so maybe it can/should be simplified even more.

Reply:
Yep, will do that ✔️ 

No, not always ❌ 

😆